i.smap: fix possible pole error with log in extract function

ymdatta opened the pull request:
Using logarithm function call with zero argument will lead to a pole error, which occurs if the mathematical function has an exact infinite result.

Refactor the conditional to only execute the code when number of subclasses are more than 1, which would preemptively stop us from getting into situation where log would have a zero argument. In cases where number of subclasses are <= 0, set the default value to 0. This is necessary as `ll` points to malloc'd memory and can contain renadom value.

References:

[1] https://en.cppreference.com/w/c/numeric/math/log
[2] [SEI CERT C Coding Standard](https://wiki.sei.cmu.edu/confluence/display/c/FLP32-C.+Prevent+or+detect+domain+and+range+errors+in+math+functions)

wenzeslaus commented on imagery/i.smap/model.c:
```diff
@@ -155,7 +155,8 @@ void extract(DCELL ***img,          /* multispectral image, img[band][i][j] */
                             subsum += exp(subll[k] - maxlike) * C->SubSig[k].pi;
 
-                        ll[i][j][m] = log(subsum) + maxlike;
+                        if (subsum > 0)
+                            ll[i][j][m] = log(subsum) + maxlike;
```
What will the be the value of ` ll[i][j][m]` in case `subsum <= 0`?

ymdatta replied:
I checked the hierarchy, `ll` points to memory allocated via malloc. So, I think it would be some random value.

We can zero this pixel value, the way we do it when we have no data at this pixel in [L117](https://github.com/OSGeo/grass/blob/main/imagery/i.smap/model.c#L117). I think that's more clean aswell.

nilason commented:
I gather this issue originates from `cppcheck`, this is the relevant part for this:

```
cppcheck imagery/i.smap/model.c
...
imagery/i.smap/model.c:158:43: warning: Invalid log() argument nr 1. The value is 0 but the valid values are '4.94066e-324:'. [invalidFunctionArg]
                        ll[i][j][m] = log(subsum) + maxlike;
                                          ^
imagery/i.smap/model.c:154:34: note: Assignment 'subsum=0', assigned value is 0
                        subsum = 0;
                                 ^
imagery/i.smap/model.c:155:39: note: Assuming condition is false
                        for (k = 0; k < C->nsubclasses; k++)
                                      ^
imagery/i.smap/model.c:158:43: note: Invalid argument
                        ll[i][j][m] = log(subsum) + maxlike;
```

The static analysis, takes the path where `C->nsubclasses` are `0`, which leaves `subsum` with the initialised value of `0.0`. `log(0)` return "_-infinity and raise the "divide-by-zero" floating-point exception_"[^1], which is the real issue here. 

Now, in this branch of the if-else statement:

https://github.com/OSGeo/grass/blob/bc9cb2a3f9540b031a6b8e8e7d7257ca01492d3e/imagery/i.smap/model.c#L144-L160

there are two loops over `C->nsubclasses` which only does anything if `C->nsubclasses > 1`.

Wouldn't it better to change the `else` to a `else if (C->nsubclasses > 1)` on line 144, which guarantees that the for-loops are entered and the initialised value of `subsum` is altered?

[^1]: `man 3 log`

ymdatta commented:
@nilason : I agree with your suggestion, it's a much cleaner way. Thanks!

I have modified the code to reflect that and I also added an `else` condition, which should cover the cases where `nsubclasses <= 0` to set the value for `ll` to `0.0`.

echoix commented:
@jayneel-shah18 @marisn Could this be one of the problems observed in i.smap?

echoix commented:
Updated since there are some tests now for i.smap.
